### ISSUE TITLE
Introduce example kubernetes deployment & documentation

### ISSUE THREAD

**Issue by Kamilczak020** [oauth2-proxy/oauth2-proxy#608]
Currently, there is no documentation and/or examples on how to deploy oauth2-proxy on kubernetes.
This would be a neat addition, especially for begineer k8s users to have something to base off of.

I'd gladly work on both the example manifests and documentation, provided it's a desired feature.

**Comment by JoelSpeed** [comment]
@evgenigordeev and I were discussing doing an example of a K8s installation in https://github.com/oauth2-proxy/oauth2-proxy/pull/604#issuecomment-640255333, perhaps you could colloborate on getting proper documentation and an example? 

I would love to have to decent documentation and examples of K8s manifests for the project

**Comment by Kamilczak020** [comment]
My idea for this would be to:

1. Introduce example manifests required for full setup of the oauth2_proxy
You could have a sneak peek into my current setup here:
https://github.com/Kamilczak020/ansible-roles/tree/master/roles/security/oauth2-proxy

It would just be a no-ansible solution of course, pure manifests, and likely a statefulset instead of a deployment.

2. Introduce documentation explaining each part of the deployment involved, with highlighting the caveats and full coverage of all the available options.

My question for this, is how in-depth would you like this to be?
Would you rather see configmaps mounted into the container with config files, or rather an argument-based solution that is easier to understand for "less-advanced" developers?

**Comment by kfox1111** [comment]
There is a helm chart available. I typically use that to get it loaded into our clusters quickly/easily. Would that be a good place to start?

**Comment by Kamilczak020** [comment]
I personally dislike helm, and I prefer not to use it - but that's preference.
But when it comes to creating a documentation about a k8s deployment, I think it would be more valuable to resort to plain ol' manifests, which are easily readable and configurable.
The helm chart could be added as an additional section in the documentation, but since it hides implementation - I believe it should not be the main focus.

**Comment by Kamilczak020** [comment]
@JoelSpeed Is there any specific way in which you'd like this to look?

**Comment by JoelSpeed** [comment]
Personally, I've never really used helm all that much, so I would not really want to be responsible for maintaining a helm chart. That said, as @kfox1111 says, there is already one available, so we should probably document its availability.

I think I would prefer plain K8s manifests, I agree that they are easily readable and I think that they lower the barrier by not requiring the use of an external dependency such as helm

> It would just be a no-ansible solution of course, pure manifests, and likely a statefulset instead of a deployment.

Before you start making a PR, why would you use a statefulset? I don't think a statefulset is required, OAuth2 Proxy should be horizontally scalable and stateless

**Comment by EvgeniGordeev** [comment]
@JoelSpeed here is a helm example https://github.com/oauth2-proxy/oauth2-proxy/pull/615 utilizing the official [oauth2-proxy](https://github.com/helm/charts/tree/master/stable/oauth2-proxy) helm chart. It could've been smaller if httpbin had a helm chart

Although helm is an extra dependency but for the sake of quick bootstrapping an example I don't see how a lengthy K8s manifest is easier as opposed to [values.yaml](https://github.com/oauth2-proxy/oauth2-proxy/pull/615/files#diff-a50a45de264546e7cd2b14ae60f9de52R1) which contains the whole configuration relevant to OIDC and oath2-proxy without going into details of the underlying infrastructure. 

I can help with supporting this helm example going forward. If you decide not to have a helm example you can get the whole manifest with `helm get manifest oauth2-proxy-example`. Or we can take the approach nginx-ingress has - publishing a manifest in automated fashion from a helm chart. 

Check this out: 
https://kind.sigs.k8s.io/docs/user/ingress/#ingress-nginx which is actually based on a helm chart https://github.com/kubernetes/ingress-nginx/tree/master/charts/ingress-nginx and the manifest is generated in automated fashion in https://github.com/kubernetes/ingress-nginx/blob/master/hack/generate-deploy-scripts.sh#L131.

**Comment by JoelSpeed** [comment]
@EvgeniGordeev I'll take a proper look at the PR hopefully later today, thanks for working on that!

I guess one of my concerns with helm previously was how hard it was to configure things by using values files and having to create really complex templates. It looks like this might be easier in the example you've created (I had a quick scan) to update values as you're just setting up a config file, +1 to that

Helping to support the helm charts sounds great to me :)

**Comment by maxbrunet** [comment]
Hey there, I have suggested a PR for Jsonnet and Kubernetes manifests/kustomize: #706. Hope this help.

**Comment by JoelSpeed** [comment]
@maxbrunet Since @EvgeniGordeev added a helm based example in #615, I'm a little hesitant to have multiple disconnected deployment examples within our repo. Do you think you could expand on why you think it's beneficial for us to have helm, raw K8s, jsonnet and kustomize examples?

**Comment by maxbrunet** [comment]
@JoelSpeed The Jsonnet/Kustomize deployments are more than just examples, they can directly be used as libraries to deploy oauth2-proxy with very little to do to get it production ready. There is not a lot of those for oauth2-proxy, the most popular, [stable/oauth2-proxy](https://github.com/helm/charts/tree/master/stable/oauth2-proxy) (not many others of quality), has been deprecated with the release of Helm 3. In my opinion, it is good to have such deployments maintain within the project.

Helm charts need to have releases to be used by others, but it does not make sense to release this charts since it is clearly an example that cannot be used as is. Kustomize and Jsonnet are good tools since nearly everything can be overridden by the end user, whereas Helm is limited.

I also see Jsonnet being useful in the future, once #580 is implemented, to build Grafana dashboards and Prometheus alert rules libraries.

This is all very similar to what is being done by the RedHat/CoreOS/OpenShift observability folks for Prometheus, Thanos, Grafana, and other Prometheus exporters.

**Comment by SayakMukhopadhyay** [comment]
I personally prefer Helm since Helm has more proliferation. It helps in having a single tool but I can also see how kustomize can help as you would not need to maintain a chart management system. At the end of the day, it would help if a quick way to deploy to k8s is available and helm and kustomize both fit the bill so its upto the devs' preferences.

**Comment by JoelSpeed** [comment]
So with the changes that are being made to Helm and the charts repo being shut down, I agree, we are going to need to host something ourselves. Personally I think Kustomize is the way to go. I'm not a fan of Helm charts as they are limited and the "release" process is frustrating, as suggested by @maxbrunet.

I'm not sure about the jsonnet stuff, to me, jsonnet makes for an overly complicated file that is hard to read and understand. I would be concerned about maintaining a jsonnet file, where something like Kustomize seems much easier to maintain.

If we do go ahead with your PR @maxbrunet, I think it would be good to update the local environment K8s example to use the Kustomize deployment files instead of the helm charts, WDYT?

@gargath @NickMeves @steakunderscore I'd appreciate your thoughts on deployment manifests and the contents of the last few comments on this thread

**Comment by NickMeves** [comment]
As far as variety for options in the future - I could potentially help with maintenance of a Helm chart if we want one internal to the project. I've been using Helm myself to deploy this in a few places (my own versions, not the stable one here: https://github.com/helm/charts/tree/master/stable/oauth2-proxy)

Even though it has its warts, I agree with @SayakMukhopadhyay 's point about the market penetration of Helm compared to alternatives at the moment.

**Comment by RichiCoder1** [comment]
As an avid fan of Helm post-v3, I'm biased in that we'd strongly prefer a helm chart as that simplifies deployment and configuration for us over kustomize/manifests. Would be willing to chip in too.

**Comment by maxbrunet** [comment]
@JoelSpeed I have opened PR #747 with Kustomize only. I have quickly looked into updating the local environment, it should be doable. If you don't mind, I will raise a separate PR once this 1st one is merged, it will make reviewing easier.

**Comment by secustor** [comment]
I haven't used helm before the release of of v2, because of security concerns. 
Since then is Helm also the go to tool for application deployment on Kubernetes for me.

**Comment by JoelSpeed** [comment]
I think I'm leaning towards suggesting we create a separate repository for manifests and have that as a Helm repo. While I've never been a fan of Helm in the past, I am aware there are a lot of improvements that have come with the v3 update.

I know a number of people have chimed in on this issue suggesting they can help maintain this, by creating a separate repo I can invite those interested in manifests to the org and make them maintainers of the chart in its own repo.

I do wonder if we could set it up so that both the Helm and Kustomize camps are happy though? I was reading [this blog post](https://blog.container-solutions.com/using-helm-and-kustomize-to-build-more-declarative-kubernetes-workloads) which discusses using both in tandem. Perhaps we could have a base helm chart and use CI to enforce that the code is generated into manifests that form a basic Kustomize set of manifests from? I believe we could generate the basic resources as shown in #747 from helm right?

What would we call the repo if we go down this route?

**Comment by RichiCoder1** [comment]
> I do wonder if we could set it up so that both the Helm and Kustomize camps are happy though? I was reading this blog post which discusses using both in tandem. Perhaps we could have a base helm chart and use CI to enforce that the code is generated into manifests that form a basic Kustomize set of manifests from? I believe we could generate the basic resources as shown in #747 from helm right?

That'd be pretty neat! Would have to do some automation to keep the two in sync, but likely be worth it. Haven't see a major org do that though.

> What would we call the repo if we go down this route?

That...would a good question. Maybe `deploy` or `manifests`?

**Comment by JoelSpeed** [comment]
> That'd be pretty neat! Would have to do some automation to keep the two in sync, but likely be worth it. Haven't see a major org do that though.

Just because no ones done it yet, doesn't mean we can't 😅 

`manifests` could be a good name. I'm wondering if helm expects a certain name for the repo or not?

**Comment by RichiCoder1** [comment]
> manifests could be a good name. I'm wondering if helm expects a certain name for the repo or not?

Nope. Heml expects the helm assets to be hosted in some storage service, and an `<repository>/index.yaml` listing out charts, versions, and tar paths. So you could host it via GitHub Pages using something like <https://github.com/helm/chart-releaser-action>

**Comment by Hokwang** [comment]
I want helm chart. I think owners can migrate from stable helm chart repo, simply.
(and I am managing many tools only helm way.)

**Comment by desaintmartin** [comment]
Maintainer of the helm/charts oauth2-proxy here. I would be happy to help migrating the helm chart and continue to maintain it here.

**Comment by JoelSpeed** [comment]
@desaintmartin If I were to create a manifests repository within the GitHub org for OAuth2 Proxy and added you as an admin, could you look after a helm chart there? Would be good to have more than one owner for this though, @RichiCoder1 would you be interested in helping maintaining the chart as well? 

Are you both on board with the ideas I outlined in https://github.com/oauth2-proxy/oauth2-proxy/issues/608#issuecomment-689047152, basically having the helm chart generate static manifests into another folder and keeping those in sync as a set of base templates for Kustomize?

**Comment by NickMeves** [comment]
Does Helm need its own repo to appropriately hook into Helm Hub (?? or whatever it is now for Helm v3) with its CI/CD?

That's what I seem to see in other major project's hosted charts:

https://github.com/prometheus-community/helm-charts/blob/main/ct.yaml

**Comment by JoelSpeed** [comment]
I had assumed we would be able to keep Kustomize in the same repo, but I'm not that familiar with Helm, perhaps @desaintmartin knows better whether we can do this or not? I definitely would prefer to keep the manifests apart from the code as it will make management of the two much easier IMO

**Comment by EvgeniGordeev** [comment]
This process https://github.com/helm/chart-releaser gets more traction, e.g. I have based our company's helm repo on it - https://github.com/Conservis/helm-charts. The release process of the charts is completely automated based on github actions and `master` branch.

**Comment by RichiCoder1** [comment]
> I had assumed we would be able to keep Kustomize in the same repo, but I'm not that familiar with Helm, perhaps @desaintmartin knows better whether we can do this or not? I definitely would prefer to keep the manifests apart from the code as it will make management of the two much easier IMO

Helm and kustomize can be in the repos. The charts would just need to be in their own folder. Most other projects have a separate chart repo because it accumulates charts across multiple subprojects, and the keep chart releases separate from project releases.

**Comment by desaintmartin** [comment]
Well, AFAIK it is easier to have a separate repository and just automate it using chart-tester and chart-releaser, but I've successfully moved the [kubernetes dashboard](https://github.com/kubernetes/dashboard/tree/master/aio/deploy/helm-chart) (that I still maintain there) within the main kube dashboard repository, although with different tooling. I suppose with a bit of work we could adapt the chart-releaser to work out of the box, if not already done.

I would still maintain the chart there! Although i'd be glad not to be alone (@Miouge1, what do you think?)

Regarding helm and kustomize, those are really different tools and although i've seen people make them interact, I am not sure we should try to stick to strict compliance between the two (like requiring a new parameter to be both in kustomize and helm) even if we should try not to diverge too much.

**Comment by JoelSpeed** [comment]
> Well, AFAIK it is easier to have a separate repository and just automate it using chart-tester and chart-releaser

This sounds good, is it a lot of effort to set up? 

> I would still maintain the chart there! Although i'd be glad not to be alone (@Miouge1, what do you think?)

If I can get an ack from @desaintmartin, @Miouge1 and @RichiCoder1 that you're on board for helping with this I'll happily set up the repo and invite you all to the org to get this project started

> Regarding helm and kustomize, those are really different tools and although i've seen people make them interact, I am not sure we should try to stick to strict compliance between the two (like requiring a new parameter to be both in kustomize and helm) even if we should try not to diverge too much.

Let's start with Helm and make Kustomize a stretch goal for a later date. I'd much rather we get something going than talk about ways of solving problems for ever 😅

**Comment by tlawrie** [comment]
Hey @JoelSpeed more than happy to donate the helm chart we have been using for Auth Proxy or help maintain the new one. I am sure @costelmoraru would feel the same (my partner in crime) https://github.com/boomerang-io/charts/tree/main/bmrg-auth-proxy 

I do see the discussion above about moving the one from the helm repo that's being sunset. @desaintmartin I wonder how different the charts are and whether we can consolidate?

I also vote for a separate repo, at the moment we are using the same GitHub as a helm repo index on a separate branch. Wrote an article about how [here](https://medium.com/@tysonlawrie/utilizing-github-releases-as-a-helm-repository-1752e39ee645). It means not needing any additional repo infra etc

**Comment by JoelSpeed** [comment]
For those of you who aren't aware, I had been speaking to @tlawrie about this issue in other threads (Slack?) and they were interested in helping out too, so I reached out to double check.

I'm going to get the new repo set up over the next few days. My suggestion would be to port verbatim the existing chart from the helm repo and then see what improvements we can bring over from the boomerang chart (though I really know very little about charts so take this with a pinch of salt)

**Comment by pierluigilenoci** [comment]
@JoelSpeed I would ask you to make the new chart backward compatible with the old one, at least at an early stage, to simplify migration. 🙏

**Comment by JoelSpeed** [comment]
So I created a repo and invited @desaintmartin and @tlawrie to own that, then I was looking at what needs doing:
- Add a license to repo
  - Do we use MIT to match the project, or Apache 2 to match where the manifests are hosted presently?
  - I suspect we have to keep Apache 2 to be compliant as we move this.
- Add an owners file which points to @oauth2-proxy/manifest-maintainers to get pull request reviews assigned
- Migrate existing helm chart (do we just copy all files?)
  - Can we move it whilst keeping the git history?
- Set up some CI?
- Set up release automation? 

The first two I understand and could do, the further ones I have questions on and don't think I could do, certainly not without answers to some questions

**Comment by tlawrie** [comment]
> License

@JoelSpeed I don't know the main difference between MIT or Apache 2. I've just always used Apache and also is recommended I think by CNCF for their projects. 

> Set up some CI / Release automation.

I am happy to set it up to put the index and everything onto GitHub, it just uses a shell script to do so and marks them as GitHub releases which I always like. I also have a script that puts them into Artifactory (which requires you to hit an API to do the index sync).

Open to suggestions if you want to publish them elsewhere. I can set that up using an IBM CICD that I have access to or translate the script to Travis CI (which I also know) or we can use any. I've been thinking about playing around with Relay.sh so would have to double-check if they have an OSS free tier.

**Comment by JoelSpeed** [comment]
> @JoelSpeed I don't know the main difference between MIT or Apache 2. I've just always used Apache and also is recommended I think by CNCF for their projects.

I think lets go for Apache 2 since that's what the main helm chart is, technically if we copy that over we have to keep the license anyway

> I am happy to set it up to put the index and everything onto GitHub, it just uses a shell script to do so and marks them as GitHub releases which I always like. I also have a script that puts them into Artifactory (which requires you to hit an API to do the index sync).

I heard we can use something like github pages to serve the chart? Do you know anything about this?

I'd suggest we set up github actions for the CICD processes that we need since that's what we are moving to with the main project repoistory

**Comment by desaintmartin** [comment]
I'm sorry I could not contribute to this issue before, but let's catch up.
1/ @tlawrie What is the difference between the two charts? Could we merge them and keep the best parts ? I don't think it has to be backward compatible as long as it is not too hard to upgrade.
2/ I think I missed the invitation, is that still ok?
3/ Let me see regarding the licence change with Helm team
4/ The "chart-releaser" (https://github.com/helm/chart-releaser-action) is exactly about hosting all versions of charts within the gh-pages. Using a dedicated repo, I think it should be really easy to have this running, as well with chart-testing (https://github.com/helm/chart-testing-action) for QA.

**Comment by pierluigilenoci** [comment]
Inactive??? 😢

**Comment by NickMeves** [comment]
@desaintmartin Let us know if you need any help from @JoelSpeed or me getting your work re-hosted here.

CC: @tlawrie @costelmoraru

**Comment by morarucostel** [comment]
> CC: @tlawrie @costelmoraru

Just tagging myself, with the account I'm using, @morarucostel

**Comment by toxuin** [comment]
Hello

**Comment by desaintmartin** [comment]
Hey again,
First thing, should we add more maintainers in https://github.com/oauth2-proxy/manifests/blob/main/helm/oauth2-proxy/Chart.yaml ? I'd like to add this into the Artifact Hub but it seems I'm the only one with email/address.

**Comment by pierluigilenoci** [comment]
😢

**Comment by desaintmartin** [comment]
Since my last comment, we've set the oauth2-proxy helm chart at https://github.com/oauth2-proxy/manifests, using chart-testing / chart-releaser, published to the helm hub.

**Comment by robin-wayve** [comment]
☝🏻 It's now documented in the Installation page. 👍🏻

**Comment by desaintmartin** [comment]
Yes ! I do not have permission to close it on this repo, @JoelSpeed? ;)
